packages/nx/src/utils/cache-directory.ts

```diff
@@ -1,40 +1,106 @@
-import { join, isAbsolute } from 'path';
-import { workspaceRoot } from './workspace-root';
-import { readJsonFile } from './fileutils';
+import { isAbsolute, join } from 'path';
 import { NxJsonConfiguration } from '../config/nx-json';
+import { readJsonFile } from './fileutils';
+import { workspaceRoot } from './workspace-root';
 
-function readCacheDirectoryProperty(root: string): string | undefined {
-  try {
-    const nxJson = readJsonFile<NxJsonConfiguration>(join(root, 'nx.json'));
-    return nxJson.tasksRunnerOptions.default.options.cacheDirectory;
-  } catch {
-    return undefined;
-  }
+export interface CacheDirectorySpec {
+  envName: string;
+  propertyName: string;
+  defaultDirectory?: string;
 }
 
-function cacheDirectory(root: string, cacheDirectory: string) {
-  const cacheDirFromEnv = process.env.NX_CACHE_DIRECTORY;
-  if (cacheDirFromEnv) {
-    cacheDirectory = cacheDirFromEnv;
+/**
+ * Calculate path to cache directory
+ * The path is calculated by looking at these values in order of precedense:
+ * env[cache.envName]
+ * nxConfig.tasksRunnerOptions.default.options[cache.propertyName]
+ * cache.defaultDirectory
+ *
+ * if the resulting path from the above rule is not an absolute path
+ * it will be prefixed with root
+ *
+ * @param root : the root of the workspace
+ * @param env : the environment - e.g. process.env
+ * @param nxConfig : the content of the workspace nx.json
+ * @param cache : the specification of the cache :
+ * -              envName: the property in env that may define the cache
+ * -              propertyName: the property in
+ * -                            nxConfig.tasksRunnerOptions.default.options
+ * -                            that may define the cache
+ * -              defaultDirectory: the cache directory path if none of the
+ * -                                above defines the path
+ * @returns
+ */
+export function cacheDirectory(
```

Comment:
I don't actually think this function should be generalized to handle this.

These are 2 different settings.

`projectGraphCache` directory isn't really an option for the task runner so I don't think it should be defined like the `cacheDirectory`.

The project graph is computed even if a task runner is not needed. (`nx graph`)

Reply:
I am not sure I see the argument here - is `packages/nx/src/utils/cache-directory.ts` specifically bound to task runner ??

So `cacheDirectory` is just a method for locating property that may be defined by environment and/or `nx.json` option, and with the proposed changes to `cacheDirectory` we would get:

```
/**
 * Path to the directory where Nx stores its cache of project dependencies and
 * logs from the Nx Daemon in case it is active.
 * It is defined in order of precedence:
 * environment variable NX_PROJECT_GRAPH_CACHE_DIRECTORY
 * node_modules/.cache/nx
 *
 * if the resulting path from the above rule is not an absolute path
 * it will be prefixed with the workspace root path
 */
export const projectGraphCacheDirectory = cacheDirectory(
  workspaceRoot,
  process.env,
  nxJson,
  {
    envName: 'NX_PROJECT_GRAPH_CACHE_DIRECTORY',
    defaultDirectory: join('node_modules', '.cache', 'nx'),
  }
);
``` 

With the proposed change to `cacheDirectory` above we would have :

```
/**
 * Path to the directory where Nx stores its cache and daemon-related files.
 * It is defined in order of precedence:
 * environment variable NX_CACHE_DIRECTORY
 * nx.json tasksRunnerOptions.default.options.cacheDirectory
 * node_modules/.cache/nx
 *
 * if the resulting path from the above rule is not an absolute path
 * it will be prefixed with the workspace root path
 */
export const cacheDir = cacheDirectory(workspaceRoot, process.env, nxJson, {
  envName: 'NX_CACHE_DIRECTORY',
  propertyName: 'cacheDirectory',
  defaultDirectory: join('node_modules', '.cache', 'nx'),
});

/**
 * Path to the directory where Nx stores its cache of project dependencies and
 * logs from the Nx Daemon in case it is active.
 * It is defined in order of precedence:
 * environment variable NX_PROJECT_GRAPH_CACHE_DIRECTORY
 * node_modules/.cache/nx
 *
 * if the resulting path from the above rule is not an absolute path
 * it will be prefixed with the workspace root path
 */
export const projectGraphCacheDirectory = cacheDirectory(
  workspaceRoot,
  process.env,
  nxJson,
  {
    envName: 'NX_PROJECT_GRAPH_CACHE_DIRECTORY',
    defaultDirectory: join('node_modules', '.cache', 'nx'),
  }
);
```

if we do not use the generalized `cacheDirectory` method we would have to duplicate some code like:

```
option and if we do not use this generalized method we would have to create a method like 
```
function projectGraphDIr() {
 const directory = process.env['NX_PROJECT_GRAPH_CACHE_DIRECTORY'] ?? join('node_modules', '.cache', 'nx')
 if (directory) {
    if (isAbsolute(directory)) {
      return directory;
    } else {
      return join(root, directory);
    }
    return '';
}
```